File: quickshift/test_ley_fundamental.py
#!/usr/bin/env python3
"""
TEST: LEY FUNDAMENTAL - Validación contra endpoint /solve

Verifica que:
1. Siempre hay ≥1 solución sin filtros
2. NUNCA aparecen cursos aprobados en soluciones
3. Funciona hasta semestre 9

Uso: python3 test_ley_fundamental.py [--server http://localhost:8080]
"""

import json
import requests
import sys
from typing import List, Dict, Tuple

class TestLeyFundamental:
    def __init__(self, server_url: str = "http://127.0.0.1:8080"):
        self.server_url = server_url
        self.endpoint = f"{server_url}/solve"
        self.passed = 0
        self.failed = 0
        self.results = []

    # Cursos por semestre (basado en Malla2020.xlsx)
    CURSOS_POR_SEMESTRE = [
        # Semestre 1
        ["CBM1000", "CBM1001", "CBQ1000", "CIT1000", "FIC1000", "CBM1002"],
        # Semestre 2
        ["CBM1003", "CBF1000", "CIT1010", "CBM1005", "CBM1006", "CBF1001"],
        # Semestre 3
        ["CIT2114", "CIT2107", "CIT1011", "CBF1002", "CIT2007", "CBF1003"],
        # Semestre 4
        ["CIT2204", "CIT2108", "CIT2009", "CBM1007", "CBM1008", "CBF1004"],
        # Semestre 5
        ["CIT2205", "CII1000", "CII1001", "CII1002", "CBF1005", "CBM1009"],
        # Semestre 6
        ["CII1003", "CII1004", "CII1005", "CII1006", "CBF1006", "CBM1010"],
        # Semestre 7
        ["CII1007", "CII1008", "CII1009", "CII1010", "CBF1007", "CBM1011"],
        # Semestre 8
        ["CII1011", "CII1012", "CII1013", "CII1014", "CBF1008", "CBM1012"],
        # Semestre 9
        ["CII1015", "CII1016", "CII1017", "CII1018", "CBF1009", "CBM1013"],
    ]

    def test_ley_fundamental_completa(self) -> bool:
        """
        Itera por semestres 1-9 aprobando cursos uno por uno.
        AMPLIADO: Cada curso se prueba MÍNIMO 3 veces en diferentes contextos:
        - Context 1: Aprobación inicial (solo ese curso)
        - Context 2: Con todos los cursos anteriores de su semestre
        - Context 3: Con el semestre completo anterior más cursos actuales
        """
        print("\n" + "="*70)
        print("🔬 TEST: LEY FUNDAMENTAL - Iteración por semestres (3+ contextos)")
        print("="*70)

        ramos_aprobados = []
        contador_total = 0

        for sem_idx, cursos_sem in enumerate(self.CURSOS_POR_SEMESTRE):
            semestre = sem_idx + 1
            print(f"\n📚 SEMESTRE {semestre}")
            print(f"   Cursos disponibles: {len(cursos_sem)}")

            for idx, curso in enumerate(cursos_sem):
                # Agregar el curso a los aprobados (persistente)
                ramos_aprobados.append(curso)
                contador_total += 1

                print(f"\n   ✓ Aprobado: {curso} ({idx+1}/{len(cursos_sem)})")
                print(f"     Total aprobados: {len(ramos_aprobados)}")

                # CONTEXTO 1: Solo cursos hasta este punto
                test_passed_1 = self._test_caso_individual(
                    semestre=semestre,
                    ramos_aprobados=ramos_aprobados.copy(),
                    idx_en_semestre=idx + 1,
                    contexto=1
                )
                if test_passed_1:
                    self.passed += 1
                else:
                    self.failed += 1

                # CONTEXTO 2: Simular que el estudiante aprobó además 1 curso aleatorio de otro semestre
                ramos_con_extra = ramos_aprobados.copy()
                if sem_idx > 0:
                    ramos_con_extra.append(self.CURSOS_POR_SEMESTRE[sem_idx - 1][0])
                    test_passed_2 = self._test_caso_individual(
                        semestre=semestre,
                        ramos_aprobados=ramos_con_extra,
                        idx_en_semestre=idx + 1,
                        contexto=2
                    )
                    if test_passed_2:
                        self.passed += 1
                    else:
                        self.failed += 1

                # CONTEXTO 3: Simular que el estudiante aprobó además 2 cursos más
                ramos_con_mas_extra = ramos_aprobados.copy()
                if sem_idx > 1:
                    ramos_con_mas_extra.extend([
                        self.CURSOS_POR_SEMESTRE[sem_idx - 2][0],
                        self.CURSOS_POR_SEMESTRE[sem_idx - 1][1]
                    ])
                    test_passed_3 = self._test_caso_individual(
                        semestre=semestre,
                        ramos_aprobados=ramos_con_mas_extra,
                        idx_en_semestre=idx + 1,
                        contexto=3
                    )
                    if test_passed_3:
                        self.passed += 1
                    else:
                        self.failed += 1

        # Resumen
        print("\n" + "="*70)
        print("\n📊 RESUMEN DEL TEST\n")
        total_casos = self.passed + self.failed
        print(f"Total de casos: {total_casos} (54 cursos × 3 contextos = 162 mínimo)")
        print(f"✅ Passed: {self.passed}")
        print(f"❌ Failed: {self.failed}")
        tasa_exito = (self.passed*100)//(total_casos) if total_casos > 0 else 0
        print(f"\n📈 Tasa de éxito: {tasa_exito}%")

        # Mostrar fallos
        if self.failed > 0:
            print(f"\n⚠️  FALLOS DETECTADOS:\n")
            for result in self.results:
                if not result["passed"]:
                    print(f"  ❌ {result['test_name']}")
                    print(f"     Razón: {result['reason']}")

        print("\n" + "="*70)
        return self.failed == 0

    def _test_caso_individual(self, semestre: int, ramos_aprobados: List[str], idx_en_semestre: int, contexto: int = 1) -> bool:
        """Ejecuta un caso individual contra /solve"""
        try:
            payload = {
                "email": "test@x.com",
                "malla": "Malla2020.xlsx",
                "sheet": "Malla 2020",
                "ramos_pasados": ramos_aprobados,
                "ramos_prioritarios": [],
                "horarios_preferidos": [],
                "filtros": {}
            }

            response = requests.post(self.endpoint, json=payload, timeout=20)
            response.raise_for_status()
            data = response.json()

            soluciones_count = data.get("soluciones_count", 0)
            soluciones = data.get("soluciones", [])

            # VALIDACIÓN 1: ¿Hay al menos 1 solución?
            if soluciones_count == 0 and len(ramos_aprobados) < len(set(c for sem in self.CURSOS_POR_SEMESTRE for c in sem)):
                test_name = f"Semestre {semestre} - {idx_en_semestre}/6 [CTX{contexto}]"
                self.results.append({
                    "test_name": test_name,
                    "passed": False,
                    "reason": f"LEY VIOLADA: 0 soluciones con {len(ramos_aprobados)} cursos aprobados"
                })
                print(f"     ❌ Contexto {contexto}: LEY VIOLADA: Sin soluciones")
                return False

            # VALIDACIÓN 2: ¿Hay cursos aprobados en las soluciones?
            ramos_set = set(ramos_aprobados)
            for sol_idx, solucion in enumerate(soluciones):
                codigos_en_sol = [sec["codigo"] for sec in solucion["secciones"]]
                aprobados_en_sol = [c for c in codigos_en_sol if c in ramos_set]

                if aprobados_en_sol:
                    test_name = f"Semestre {semestre} - {idx_en_semestre}/6 [CTX{contexto}]"
                    self.results.append({
                        "test_name": test_name,
                        "passed": False,
                        "reason": f"Cursos aprobados en solución: {aprobados_en_sol}"
                    })
                    print(f"     ❌ Contexto {contexto}: Cursos aprobados en solución: {aprobados_en_sol}")
                    return False

            # VALIDACIÓN 3: Contar soluciones válidas
            test_name = f"Semestre {semestre} - {idx_en_semestre}/6 [CTX{contexto}]"
            self.results.append({
                "test_name": test_name,
                "passed": True,
                "reason": f"✅ {soluciones_count} soluciones válidas"
            })
            print(f"     ✅ Contexto {contexto}: {soluciones_count} soluciones válidas (sin aprobados)")
            return True

        except Exception as e:
            test_name = f"Semestre {semestre} - {idx_en_semestre}/6 [CTX{contexto}]"
            self.results.append({
                "test_name": test_name,
                "passed": False,
                "reason": f"ERROR: {str(e)}"
            })
            print(f"     ❌ Contexto {contexto}: ERROR: {str(e)}")
            return False

    def test_sin_filtros_garantia(self) -> bool:
        """Verifica que SIN FILTROS siempre hay solución"""
        print("\n" + "="*70)
        print("🔬 TEST: Garantía - Sin filtros = siempre solución")
        print("="*70)

        for sem_idx in range(len(self.CURSOS_POR_SEMESTRE) - 1):  # No el último
            ramos_aprobados = []
            for i in range(sem_idx + 1):
                ramos_aprobados.extend(self.CURSOS_POR_SEMESTRE[i])

            semestre = sem_idx + 1

            try:
                payload = {
                    "email": "test@x.com",
                    "malla": "Malla2020.xlsx",
                    "sheet": "Malla 2020",
                    "ramos_pasados": ramos_aprobados,
                    "ramos_prioritarios": [],
                    "horarios_preferidos": [],
                    "filtros": {}
                }

                response = requests.post(self.endpoint, json=payload, timeout=20)
                data = response.json()
                soluciones_count = data.get("soluciones_count", 0)

                if soluciones_count > 0:
                    print(f"✅ Semestre {semestre+1}: {soluciones_count} soluciones (LEY cumplida)")
                    self.passed += 1
                else:
                    print(f"❌ Semestre {semestre+1}: 0 soluciones (LEY VIOLADA)")
                    self.failed += 1

            except Exception as e:
                print(f"❌ Semestre {semestre+1}: ERROR - {str(e)}")
                self.failed += 1

        print(f"\n✅ Passed: {self.passed} | ❌ Failed: {self.failed}")
        return self.failed == 0

    def run_all_tests(self) -> bool:
        """Ejecuta todos los tests"""
        print("\n🚀 Iniciando validación de LEY FUNDAMENTAL\n")
        print(f"   Servidor: {self.server_url}")
        print(f"   Endpoint: {self.endpoint}\n")

        # Test 1: Iteración completa
        result1 = self.test_ley_fundamental_completa()

        # Test 2: Garantía de soluciones
        result2 = self.test_sin_filtros_garantia()

        print("\n" + "="*70)
        if result1 and result2:
            print("✅ TODOS LOS TESTS PASARON - LEY FUNDAMENTAL VERIFICADA")
        else:
            print("❌ ALGUNOS TESTS FALLARON - REVISAR LOGS")
        print("="*70 + "\n")

        return result1 and result2


if __name__ == "__main__":
    server_url = "http://127.0.0.1:8080"

    # Parsear argumentos
    if len(sys.argv) > 2 and sys.argv[1] == "--server":
        server_url = sys.argv[2]

    tester = TestLeyFundamental(server_url=server_url)
    success = tester.run_all_tests()

    sys.exit(0 if success else 1)

File: quickshift/tests/test_ley_fundamental.rs
/// TEST: Ley Fundamental - Siempre hay solución sin filtros
/// 
/// Esta prueba verifica que el sistema cumpla la LEY FUNDAMENTAL:
/// "Mientras quedan cursos por aprobar y NO hay filtros, 
///  SIEMPRE debe haber al menos 1 solución"
///
/// Metodología:
/// - Itera por semestres 1-9
/// - En cada semestre, aprueba cursos uno por uno
/// - Verifica: 1) ≥1 solución sin filtros, 2) Sin cursos aprobados en resultado

#[cfg(test)]
mod test_ley_fundamental {
    use std::collections::HashSet;

    // Estructura de datos para simular la progresión académica
    struct EstadoAcademico {
        semestre: usize,
        ramos_aprobados: Vec<String>,
        ramos_disponibles: Vec<String>,
    }

    /// Cursos por semestre (basado en Malla2020.xlsx)
    fn cursos_por_semestre() -> Vec<Vec<&'static str>> {
        vec![
            // Semestre 1
            vec!["CBM1000", "CBM1001", "CBQ1000", "CIT1000", "FIC1000", "CBM1002"],
            // Semestre 2
            vec!["CBM1003", "CBF1000", "CIT1010", "CBM1005", "CBM1006", "CBF1001"],
            // Semestre 3
            vec!["CIT2114", "CIT2107", "CIT1011", "CBF1002", "CIT2007", "CBF1003"],
            // Semestre 4
            vec!["CIT2204", "CIT2108", "CIT2009", "CBM1007", "CBM1008", "CBF1004"],
            // Semestre 5
            vec!["CIT2205", "CII1000", "CII1001", "CII1002", "CBF1005", "CBM1009"],
            // Semestre 6
            vec!["CII1003", "CII1004", "CII1005", "CII1006", "CBF1006", "CBM1010"],
            // Semestre 7
            vec!["CII1007", "CII1008", "CII1009", "CII1010", "CBF1007", "CBM1011"],
            // Semestre 8
            vec!["CII1011", "CII1012", "CII1013", "CII1014", "CBF1008", "CBM1012"],
            // Semestre 9
            vec!["CII1015", "CII1016", "CII1017", "CII1018", "CBF1009", "CBM1013"],
        ]
    }

    #[test]
    fn test_ley_fundamental_completa() {
        println!("\n🔬 TEST: LEY FUNDAMENTAL - Iteración por semestres\n");
        println!("{}", "=".repeat(60));

        let cursos_por_sem = cursos_por_semestre();
        let mut ramos_aprobados: Vec<String> = Vec::new();
        let mut test_results = Vec::new();

        for (sem_idx, cursos_sem) in cursos_por_sem.iter().enumerate() {
            let semestre = sem_idx + 1;
            println!("\n📚 SEMESTRE {}", semestre);
            println!("   Cursos disponibles: {:?}", cursos_sem);

            // En cada semestre, aprobamos cursos uno por uno
            for (idx, curso) in cursos_sem.iter().enumerate() {
                // Agregar el curso a los aprobados
                ramos_aprobados.push(curso.to_string());

                let cursos_aprobados_str = ramos_aprobados
                    .iter()
                    .map(|c| format!("\"{}\"", c))
                    .collect::<Vec<_>>()
                    .join(",");

                println!("\n   ✓ Aprobado: {} ({}/{})", curso, idx + 1, cursos_sem.len());
                println!("     Total aprobados: {}", ramos_aprobados.len());

                // Aquí iría la llamada a /solve con estos ramos_pasados
                // Por ahora solo validamos la lógica
                let test_case = format!(
                    "Semestre {} - {} cursos aprobados",
                    semestre,
                    ramos_aprobados.len()
                );

                // VALIDACIÓN 1: Verificar que hay cursos pendientes
                let cursos_pendientes = cursos_por_sem
                    .iter()
                    .flat_map(|c| c.iter())
                    .filter(|c| !ramos_aprobados.contains(&c.to_string()))
                    .count();

                if cursos_pendientes > 0 {
                    println!("     ✅ Hay {} cursos pendientes", cursos_pendientes);
                    test_results.push((test_case, true, cursos_pendientes));
                } else {
                    println!("     ⚠️  Sin cursos pendientes");
                    test_results.push((test_case, true, 0));
                }
            }
        }

        // Resumen
        println!("\n{}", "=".repeat(60));
        println!("\n📊 RESUMEN DEL TEST\n");

        let total_tests = test_results.len();
        let tests_ok = test_results.iter().filter(|(_, ok, _)| *ok).count();

        for (test_name, ok, pendientes) in &test_results {
            let status = if *ok { "✅" } else { "❌" };
            if *pendientes > 0 {
                println!("{} {} (Pendientes: {})", status, test_name, pendientes);
            } else {
                println!("{} {}", status, test_name);
            }
        }

        println!("\n{}", "=".repeat(60));
        println!("\n✅ RESULTADOS: {}/{} tests passed", tests_ok, total_tests);
        println!("\n📝 NOTAS IMPORTANTES:");
        println!("  1. Este test valida la LÓGICA de la progresión académica");
        println!("  2. Para validación completa, se requiere ejecutar contra /solve");
        println!("  3. Cada caso debe verificar:");
        println!("     • ≥1 solución SIN filtros");
        println!("     • CERO cursos aprobados en la solución");
        println!("  4. Si faltra alguna validación → BUG EN EL SISTEMA");
        println!();

        assert_eq!(
            tests_ok, total_tests,
            "{}% de tests pasaron",
            (tests_ok * 100) / total_tests
        );
    }

    /// Test que valida que ningún curso aprobado aparece en soluciones
    #[test]
    fn test_sin_cursos_aprobados_en_solucion() {
        println!("\n🔬 TEST: Validación - Sin cursos aprobados en soluciones\n");

        let cursos_por_sem = cursos_por_semestre();
        let mut ramos_aprobados = HashSet::new();

        // Simular aprobar cursos del semestre 1
        for curso in cursos_por_sem[0].iter() {
            ramos_aprobados.insert(curso.to_string());
        }

        println!("Cursos aprobados: {:?}", ramos_aprobados);

        // Aquí se hace la llamada a /solve con ramos_pasados
        // y se valida que NO aparezcan en la solución
        let soluciones_esperadas = vec!["CIT2114", "CIT2107", "CIT1011", "CBF1002"];

        println!("Soluciones esperadas (semestre 2+): {:?}", soluciones_esperadas);

        // Verificar que ningún curso esperado está en aprobados
        for curso in soluciones_esperadas.iter() {
            assert!(
                !ramos_aprobados.contains(&curso.to_string()),
                "VIOLACIÓN: {} está en aprobados pero en solución",
                curso
            );
            println!("  ✅ {} NO está en aprobados", curso);
        }

        println!("\n✅ Test completado: Sin violaciones");
    }

    /// Test que valida la progresión hasta semestre 9
    #[test]
    fn test_progresion_hasta_semestre_9() {
        println!("\n🔬 TEST: Progresión completa hasta Semestre 9\n");

        let cursos_por_sem = cursos_por_semestre();
        let mut ramos_aprobados = Vec::new();
        let mut contador = 0;

        for (sem_idx, cursos_sem) in cursos_por_sem.iter().enumerate() {
            let semestre = sem_idx + 1;

            for curso in cursos_sem.iter() {
                ramos_aprobados.push(curso.to_string());
                contador += 1;

                if contador % 6 == 0 {
                    // Cada 6 cursos (1 semestre completo)
                    println!("✅ Semestre {} completado | Total: {}", semestre, contador);
                }
            }
        }

        println!("\n🎓 PROGRAMA COMPLETADO");
        println!("   Total de cursos aprobados: {}", contador);
        println!("   Semestres completados: {}", cursos_por_sem.len());

        assert_eq!(contador, 54, "Debe haber 54 cursos en total (9 semestres × 6)");
        println!("\n✅ Test completado: Estructura de malla validada");
    }

    /// TEST: Validar que los filtros de horarios funcionan correctamente
    /// Aplica restricciones de horario aleatorias y verifica que:
    /// 1. Se pueden aplicar filtros sin errores
    /// 2. El número de soluciones con filtros es ≤ sin filtros
    #[test]
    fn test_filtros_horarios_funcionan() {
        use quickshift::api_json::InputParams;
        use quickshift::algorithm::ejecutar_ruta_critica_with_params;
        use quickshift::models::{UserFilters, DiaHorariosLibres};

        println!("\n🧪 TEST: Validación de Filtros de Horarios\n");
        println!("{}", "=".repeat(70));

        // Parámetros base: estudiante con algunos cursos aprobados
        let ramos_aprobados = vec![
            "CBM1000".to_string(),
            "CBM1001".to_string(),
            "CBQ1000".to_string(),
        ];

        // TEST 1: SIN FILTROS
        println!("\n📋 TEST 1: Ejecutar SIN filtros");
        let params_sin_filtros = InputParams {
            email: "test@example.com".to_string(),
            ramos_pasados: ramos_aprobados.clone(),
            ramos_prioritarios: vec![],
            horarios_preferidos: vec![],
            malla: "MiMalla.xlsx".to_string(),
            anio: None,
            sheet: None,
            student_ranking: Some(0.75),
            ranking: None,
            filtros: None,
        };

        let soluciones_sin_filtros = match ejecutar_ruta_critica_with_params(params_sin_filtros) {
            Ok(sol) => sol,
            Err(e) => {
                eprintln!("❌ Error ejecutando sin filtros: {}", e);
                vec![]
            }
        };

        println!("✅ Soluciones SIN filtros: {}", soluciones_sin_filtros.len());
        if !soluciones_sin_filtros.is_empty() {
            println!("   Primera solución: {} cursos", soluciones_sin_filtros[0].0.len());
            for (sec, _) in soluciones_sin_filtros[0].0.iter().take(3) {
                println!("     - {} (Horario: {})", sec.codigo, sec.horario.join(" | "));
            }
        }

        // TEST 2: CON FILTRO DE HORARIO - Excluir mañanas (08:00-12:00)
        println!("\n📋 TEST 2: Ejecutar CON filtro de horario (sin 08:00-12:00)");
        let mut filtros_con_restriccion = UserFilters::default();
        filtros_con_restriccion.dias_horarios_libres = Some(DiaHorariosLibres {
            habilitado: true,
            dias_libres_preferidos: None,
            minimizar_ventanas: None,
            ventana_ideal_minutos: None,
            franjas_prohibidas: Some(vec![
                "LU 08:00-12:00".to_string(),
                "MA 08:00-12:00".to_string(),
                "MI 08:00-12:00".to_string(),
                "JU 08:00-12:00".to_string(),
                "VI 08:00-12:00".to_string(),
            ]),
            no_sin_horario: Some(false),
        });

        let params_con_filtros = InputParams {
            email: "test@example.com".to_string(),
            ramos_pasados: ramos_aprobados.clone(),
            ramos_prioritarios: vec![],
            horarios_preferidos: vec![],
            malla: "MiMalla.xlsx".to_string(),
            anio: None,
            sheet: None,
            student_ranking: Some(0.75),
            ranking: None,
            filtros: Some(filtros_con_restriccion),
        };

        let soluciones_con_filtros = match ejecutar_ruta_critica_with_params(params_con_filtros) {
            Ok(sol) => sol,
            Err(e) => {
                eprintln!("❌ Error ejecutando con filtros: {}", e);
                vec![]
            }
        };

        println!("✅ Soluciones CON filtros: {}", soluciones_con_filtros.len());
        if !soluciones_con_filtros.is_empty() {
            println!("   Primera solución: {} cursos", soluciones_con_filtros[0].0.len());
            for (sec, _) in soluciones_con_filtros[0].0.iter().take(3) {
                println!("     - {} (Horario: {})", sec.codigo, sec.horario.join(" | "));
            }
        }

        // VALIDACIONES
        println!("\n{}", "=".repeat(70));
        println!("\n📊 ANÁLISIS DE RESULTADOS:\n");

        let tiene_soluciones_sin_filtros = !soluciones_sin_filtros.is_empty();
        let tiene_soluciones_con_filtros = !soluciones_con_filtros.is_empty();

        println!("✓ Soluciones sin filtros: {}", soluciones_sin_filtros.len());
        println!("✓ Soluciones con filtros: {}", soluciones_con_filtros.len());

        if tiene_soluciones_sin_filtros {
            println!("✅ RESULTADO 1: Hay soluciones sin filtros");
        } else {
            println!("⚠️  RESULTADO 1: NO hay soluciones sin filtros (esperado >0)");
        }

        if tiene_soluciones_con_filtros {
            println!("✅ RESULTADO 2: Hay soluciones con filtros");
            
            // Validar que los cursos tienen horarios válidos (no en franja prohibida)
            let mut todos_validos = true;
            for (sol, _) in soluciones_con_filtros.iter() {
                for (sec, _) in sol.iter() {
                    for horario in &sec.horario {
                        // Verificar que NO contiene "08:00-12:00" (esto es simplificado)
                        if horario.contains("08:") && (horario.contains("09:") || horario.contains("10:") || horario.contains("11:")) {
                            println!("⚠️  Advertencia: {} tiene horario potencialmente conflictivo: {}", sec.codigo, horario);
                            todos_validos = false;
                        }
                    }
                }
            }
            
            if todos_validos {
                println!("✅ RESULTADO 3: Todos los horarios evitan la franja prohibida");
            } else {
                println!("⚠️  RESULTADO 3: Algunos horarios podrían violar el filtro");
            }
        } else {
            println!("⚠️  RESULTADO 2: NO hay soluciones con filtros (puede ser válido si filtro es muy restrictivo)");
        }

        // Comparación
        let relacion = if soluciones_sin_filtros.is_empty() {
            "N/A".to_string()
        } else {
            format!("{}%", (soluciones_con_filtros.len() * 100) / soluciones_sin_filtros.len())
        };

        println!("\n📈 Relación (con/sin filtros): {}", relacion);

        // CONCLUSIÓN
        println!("\n{}", "=".repeat(70));
        if tiene_soluciones_sin_filtros {
            println!("✅ TEST PASSED: Filtros funcionan (sin errores)");
        } else {
            println!("❌ TEST FAILED: No hay soluciones base");
        }
        println!();

        assert!(
            tiene_soluciones_sin_filtros,
            "Debe haber ≥1 solución sin filtros"
        );
    }

    /// TEST: Filtros restrictivos múltiples
    /// Aplica varios filtros simultáneamente
    #[test]
    fn test_filtros_multiples_simultaneos() {
        use quickshift::api_json::InputParams;
        use quickshift::algorithm::ejecutar_ruta_critica_with_params;
        use quickshift::models::{UserFilters, DiaHorariosLibres, PreferenciasProfesores};

        println!("\n🧪 TEST: Filtros Múltiples Simultáneos\n");
        println!("{}", "=".repeat(70));

        let ramos_aprobados = vec![
            "CBM1000".to_string(),
            "CBM1001".to_string(),
            "CBQ1000".to_string(),
            "CIT1000".to_string(),
        ];

        // Crear filtros múltiples
        let mut filtros = UserFilters::default();
        
        // Filtro 1: Horarios
        filtros.dias_horarios_libres = Some(DiaHorariosLibres {
            habilitado: true,
            dias_libres_preferidos: None,
            minimizar_ventanas: None,
            ventana_ideal_minutos: None,
            franjas_prohibidas: Some(vec![
                "VI 08:00-18:00".to_string(), // Sin clases los viernes
            ]),
            no_sin_horario: Some(false),
        });

        // Filtro 2: Profesores
        filtros.preferencias_profesores = Some(PreferenciasProfesores {
            habilitado: false, // Deshabilitado para no restringir tanto
            profesores_preferidos: None,
            profesores_evitar: None,
        });

        let params = InputParams {
            email: "test@example.com".to_string(),
            ramos_pasados: ramos_aprobados,
            ramos_prioritarios: vec![],
            horarios_preferidos: vec![],
            malla: "MiMalla.xlsx".to_string(),
            anio: None,
            sheet: None,
            student_ranking: Some(0.75),
            ranking: None,
            filtros: Some(filtros),
        };

        println!("📋 Parámetros:");
        println!("   - Filtro de horarios: SÍ (sin viernes)");
        println!("   - Filtro de profesores: NO");
        println!("   - Ramos aprobados: 4");

        let soluciones = match ejecutar_ruta_critica_with_params(params) {
            Ok(sol) => sol,
            Err(e) => {
                eprintln!("❌ Error: {}", e);
                vec![]
            }
        };

        println!("\n✅ Soluciones obtenidas: {}", soluciones.len());

        if !soluciones.is_empty() {
            for (idx, (sol, score)) in soluciones.iter().take(2).enumerate() {
                println!("\n📌 Solución #{}: score={}", idx + 1, score);
                for (sec, _) in sol.iter() {
                    println!("   - {} (Día: {})", sec.codigo, sec.horario.join(", "));
                }
            }
        }

        println!("\n✅ TEST PASSED: Filtros múltiples se aplican sin errores\n");
        assert!(soluciones.len() > 0, "Debe haber al menos 1 solución");
    }

}

/*
Este test puede ejecutarse con:
cargo test --test test_ley_fundamental -- --nocapture

OUTPUT ESPERADO:

🔬 TEST: LEY FUNDAMENTAL - Iteración por semestres

============================================================

📚 SEMESTRE 1
   Cursos disponibles: ["CBM1000", "CBM1001", "CBQ1000", "CIT1000", "FIC1000", "CBM1002"]

   ✓ Aprobado: CBM1000 (1/6)
     Total aprobados: 1
     ✅ Hay 53 cursos pendientes

   ✓ Aprobado: CBM1001 (2/6)
     ...

📊 RESUMEN DEL TEST

✅ RESULTADOS: 54/54 tests passed

📝 NOTAS IMPORTANTES:
  1. Este test valida la LÓGICA de la progresión académica
  2. Para validación completa, se requiere ejecutar contra /solve
  3. Cada caso debe verificar:
     • ≥1 solución SIN filtros
     • CERO cursos aprobados en la solución
  4. Si falta alguna validación → BUG EN EL SISTEMA
*/
